packages/apps/job-launcher/server/src/common/interceptors/enum-insensitive.ts

```diff
@@ -0,0 +1,35 @@
+import {
+  Injectable,
+  NestInterceptor,
+  ExecutionContext,
+  CallHandler,
+} from '@nestjs/common';
+import { Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+
+@Injectable()
+export class EnumTransformInterceptor implements NestInterceptor {
+  intercept(context: ExecutionContext, next: CallHandler): Observable<any> {
+    const request = context.switchToHttp().getRequest();
+
+    this.transformEnums(request.body);
+
+    return next.handle().pipe(
+      map((data) => {
+        return data;
+      }),
+    );
+  }
+
+  private transformEnums(obj: any) {
+    if (typeof obj !== 'object' || obj === null) return obj;
+
+    for (const key of Object.keys(obj)) {
+      if (typeof obj[key] === 'string') {
```

Comment:
Using interceptor, it is impossible to determine whether a property belongs to the enum type or not by internal means, since all we have is a body. We can come up with some clever mapping, when we will map each request and its body and determine its belonging to the internal server type.
```json
{
  chainId: 80002,
  qualifications: [],
  requesterTitle: 'Plant Phenotyping Data Collection',
  requesterDescription: 'A task to gather 3D plant models and multispectral data for precision agriculture analysis.',
  submissionsRequired: 100,
  fundAmount: 1,
  currency: 'HMT' // string
}
```

Reply:
So we can try to implement mapping, or use validators for each property in the dto, or try to use them globally.

Reply:
Here is an example
```typescript
import {
    Injectable,
    NestInterceptor,
    ExecutionContext,
    CallHandler,
  } from '@nestjs/common';
  import { Observable } from 'rxjs';
  import { map } from 'rxjs/operators';
import { JobCurrency } from '../enums/job';
  
  @Injectable()
  export class EnumTransformInterceptor implements NestInterceptor {
    intercept(context: ExecutionContext, next: CallHandler): Observable<any> {
      const request = context.switchToHttp().getRequest();
  
      this.transformEnums(request.body);
  
      return next.handle().pipe(
        map((data) => {
          return data;
        }),
      );
    }
  
    private transformEnums(obj: any) {
      if (typeof obj !== 'object' || obj === null) return obj;
  
      for (const key of Object.keys(obj)) {
        if (key === 'currency' && typeof obj[key] === 'string') {
          obj[key] = this.mapCurrencyEnum(obj[key]);
        } else if (typeof obj[key] === 'object') {
          this.transformEnums(obj[key]);
        }
      }
    }
  
    private mapCurrencyEnum(value: string): JobCurrency {
      const lowerValue = value.toLowerCase();
  
      switch (lowerValue) {
        case JobCurrency.HMT.toLowerCase():
          return JobCurrency.HMT;
        case JobCurrency.USD.toLowerCase():
          return JobCurrency.USD;
        default:
          throw new Error(`Invalid currency value: ${value}`);
      }
    }
  }
```